README.md
---
title: "README"
author: "stanislav zamecnik"
date: "15 3 2021"
output:
  html_document: default
  pdf_document: default
---
# OEFPIL


## Optimal Estimation of Parameters by Iterated Linearization

The original version of this software was written in R by Stanislav Zamecnik, Zdenka Gerslova and Vojtech Sindlar in year 2021. The package is based on theoretical background of work of prof. Gejza Wimmer and afterwards implemented by mentioned authors. 
Main features of package include:

- estimation of parameters of nonlinear function by iterated linearization
- possibility to use generic functions to OEFPIL objects
  - extract confidence bands for set of points
  - confidence intervals for parameters 
  - extract summary of used model
  - get covariance matrix for model parameters.
- plot the OEFPIL object, when one is created in a different ways
  - plot of estimated curve 
  - plot of estimated curve with ggplot2 package
- count orthogonal residuals for OEFPIL object
- print out information about OEFPIL object
- calculate estimates of parameters in Nanoindentation
- two datasets were added to the package to show functionality

## Installation

You can install release from Github repository by:

``` r
devtools::install_github("stazam/OEFPIL-")
``` 
or you install this release from [CRAN](https://CRAN.R-project.org) with: 

``` r
install.packages("OEFPIL")
``` 

## Usage

In R session do:


```r
library(MASS)
steamdata <- steam
colnames(steamdata) <- c("x","y")
k <- nrow(steamdata)
CM <- diag(rep(10,2*k))
```

Creating OEFPIL object which we want to work with


```r
library(OEFPIL)
st1 <- OEFPIL(steamdata, y ~ b1 * 10 ^ (b2 * x/ (b3 + x)),
list(b1 = 5, b2 = 8, b3 = 200), CM, useNLS = FALSE)
```

Displaying results using summary function

```r
summary(st1)
```

```
## Summary of the result:  
##  
## y ~ b1 * 10^(b2 * x/(b3 + x))
## 
##     Param Est         Std Dev   CI Bound 2.5 %   CI Bound 97.5 %
## b1   4.487870        1.526903         1.495196          7.480545
## b2   7.188155        1.865953         3.530953         10.845356
## b3 221.837783       99.953658        25.932214        417.743352
## 
##  Estimated covariance matrix: 
##            b1         b2        b3
## b1   2.331432   2.296195  134.3054
## b2   2.296195   3.481782  184.6313
## b3 134.305405 184.631318 9990.7337
## 
##  Number of iterations: 10
```

Plot of estimated function

```r
plot(st1, signif.level = 0.05, interval = "conf", main  = "Estimated function by iterated linearization")
```

![](man/figure/unnamed-chunk-4-1.png)

Ggplot graph of estimated function 

```r
library(ggplot2)
curvplot.OEFPIL(st1, signif.level = 0.05)
```

![](man/figure/unnamed-chunk-5-1.png)

For more information and examples see:

```r
?OEFPIL
```
This software OEFPIL was financed by the Technology Agency of the Czech Republic within the ZETA Programme. https://www.tacr.cz . 

<p align= "center">
<img src = "man/figure/logo_TACR.png" width = "200">
</p>




NEWS.md
---
title: "NEWS"
author: "stanislav zamecnik"
date: "15 3 2021"
output:
  html_document: default
  pdf_document: default
---

## OEFPIL 0.1.0

- inital release (10.03.2021)


R/plot.OEFPIL.R
#' @name plot.OEFPIL
#' @title Plot the estimate from an OEPFIL object
#' @description Plot of the iterated linearization estimate of a function from an \code{"OEFPIL"} object with pointwise confidence and prediction bands.
#'
#' @param x an object of class \code{"OEFPIL"} (a result of a call to \code{\link{OEFPIL}}).
#' @param xx  a sequence of x-coordinates of points for computing and plotting confidence bands. If missing, the default sequence \code{seq(from = min(x), to = max(x), length.out = 301)} is used.
#' @param signif.level a numerical value or a vector of significance levels for confidence bands.
#' @param interval a character vector. It states type of an interval to draw. Following values are possible: \code{"conf"} for confidence interval, \code{"pred"} for prediction interval or \code{c("conf", "pred")} for both. If missing, no intervals are plotted.
#' @param new.obs.variance the variance of a new observation for prediction interval computing (see \code{\link{confBands.OEFPIL}}).
#' @param ... additional arguments (same as in \link{plot} function) affecting the plot.
#' @details If the \code{signif.level} argument is missing, even though an \code{interval} argument is set to \code{"conf"}, the default value 0.05 is used.
#'          The line type is set to \code{'dashed'} for confidence bands and \code{'dotted'} for prediction bands.
#'          The confidence and prediction bands are computed under normality assumption. If the vector \code{signif.level} length is greater than 1,
#'          then multiple bands are plotted. The widest band has colour no. 2. The second widest band has colour no. 3 etc.
#'
#' @return Returns an object of type list containing at least the following components
#'
#' \item{xx}{a numerical vector of points where bands are calculated.}
#' \item{yy}{a numerical vector with values of estimated function in \code{xx}.}
#' \item{PointwiseCB}{a matrix of confidence intervals at points \code{xx}.}
#' \item{PredictCB}{a matrix of prediction intervals at points \code{xx}.}
#'
#'
#' @seealso \code{\link{OEFPIL}}
#'
#' @examples
#' library(MASS)
#'
#' ##Use of plot function with default parameters
#' ##(only estimation of the function is plotted, without confidence or prediction bands)
#' steamdata <- steam
#' colnames(steamdata) <- c("x","y")
#' n <- nrow(steamdata)
#' CM1 <- diag(rep(10,2*n))
#' st1 <- OEFPIL(steamdata, y ~ b1 * 10^(b2 * x/ (b3 + x)), list(b1 = 5, b2 = 8, b3 = 200),
#'              CM1, useNLS = FALSE)
#' plot(st1)
#'
#' ##Use of plot function for plotting confidence bands
#' plot(st1, seq(0,113,0.1), signif.level = c(0.01,0.05), interval = "conf",
#'  main = "Graph of estimated function")
#'
#'##Use of plot function for plotting prediction bands
#' plot(st1, seq(0,113,0.1), interval = "pred", new.obs.variance = 15)
#'
#' ##Return values of plot function
#' (a <- plot(st1, signif.level = 0.05, interval = "conf"))
#'
#' @import graphics
#'
#' @method plot OEFPIL
#' @export
plot.OEFPIL <- function(x, xx, signif.level, interval, new.obs.variance, ...) {

  ## Function plots confidence bands of list from OEFPIL() function.
  ## x           . . . output from OEFPIL()
  ## xx          . . . in these points we calculate and plot CI (confidence intervals) or
  ##                   CB (conf. bands)
  ## interval    . . . character vector; It states type of interval to draw. It can take values:
  ##                   "conf" (confidence int.) or "pred" (prediction int.) or c("conf", "pred") (both).

  if (missing(signif.level)) {
    signif.level <- 0.05

  }

  d <- length(signif.level)

  x_1 <- x$contents[[3]] ## x_1 data
  y <- x$contents[[4]] ## y data

  dep.var.name <- x$contents$dep.var.name ## name of dependant variabe
  idp.var.name <- x$contents$idp.var.name ## name of independant variable

  if (missing(xx)) {
    xx <- seq(from = min(x_1), to = max(x_1), length.out = 301)
  }

  if (missing(new.obs.variance)) {
    CM <- x$contents$CM
    n <- length(diag(CM)) / 2
    new.obs.variance <- mean(diag(CM)[(n+1):(2*n)])
    ## "estimation" of variance of the new observation (needed for prediction interval)
  }

  CB <- confBands(x, xx = xx, signif.level = signif.level, new.obs.variance = new.obs.variance)

  plot(x_1, y, xlab = idp.var.name, ylab = dep.var.name, ... = ...)

  lines(CB$xx, CB$yy, lwd = 2, col = "black")

  if(missing(interval)){
    interval <- "none"
  }


  if (any(interval == "conf")) {
    for (i in 0:(d-1)) {
      lines(CB$xx, CB$PointwiseCB[, i+1], lwd = 2, lty = 2, col = i + 2)
      lines(CB$xx, CB$PointwiseCB[, 2*d - i], lwd = 2, lty = 2, col = i + 2)
    }
  }

  if (any(interval == "pred")) {
    for (i in 0:(d-1)) {
      lines(CB$xx, CB$PredictCB[, i+1], lwd = 2, lty = 3, col = i + 2)
        lines(CB$xx, CB$PredictCB[, 2*d - i], lwd = 2, lty = 3, col = i + 2)
    }
  }

  return(invisible(list(xx = xx, yy = CB$yy, PointwiseCB = CB$PointwiseCB,
                        PredictCB = CB$PredictCB)))
}
